### DISCUSSION TITLE
Airflow on Ubuntu 20.04 crashes each time it receives a request

### DISCUSSION THREAD

**Discussion by ijm667** [apache/airflow#35351] [General]
### Apache Airflow version

2.7.2

### What happened

I installed the recent version of airflow and it crashes when I try and access from my CHROME web browser, and get the following message:

*"Something bad has happened.

Airflow is used by many users, and it is very likely that others had similar problems and you can easily find
a solution to your problem."

The configuration used to work about 6 month ago, out of the box, but now its no longer working.

Each requests produces the following PYTHON CRASH :

```console
[^[[34m2023-11-01T14:24:42.874+0000^[[0m] {^[[34mapp.py:^[[0m1744} ERROR^[[0m - Exception on /favicon.ico [GET]^[[0m
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/flask/app.py", line 2529, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/local/lib/python3.8/dist-packages/flask/app.py", line 1826, in full_dispatch_request
    return self.finalize_request(rv)
  File "/usr/local/lib/python3.8/dist-packages/flask/app.py", line 1847, in finalize_request
    response = self.process_response(response)
  File "/usr/local/lib/python3.8/dist-packages/flask/app.py", line 2344, in process_response
    self.session_interface.save_session(self, ctx.session, response)
  File "/usr/local/lib/python3.8/dist-packages/airflow/www/session.py", line 33, in save_session
    return super().save_session(*args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/flask_session/sessions.py", line 573, in save_session
    response.set_cookie(app.config["SESSION_COOKIE_NAME"], session_id,
  File "/usr/local/lib/python3.8/dist-packages/werkzeug/sansio/response.py", line 224, in set_cookie
    dump_cookie(
  File "/usr/local/lib/python3.8/dist-packages/werkzeug/http.py", line 1303, in dump_cookie
    if not _cookie_no_quote_re.fullmatch(value):
TypeError: cannot use a string pattern on a bytes-like object
[^[[34m2023-11-01T14:24:42.887+0000^[[0m] {^[[34mapp.py:^[[0m1852} ERROR^[[0m - Request finalizing failed with an error while handling an error^[[0m
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/flask/app.py", line 2529, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/local/lib/python3.8/dist-packages/flask/app.py", line 1826, in full_dispatch_request
    return self.finalize_request(rv)
  File "/usr/local/lib/python3.8/dist-packages/flask/app.py", line 1847, in finalize_request
    response = self.process_response(response)
  File "/usr/local/lib/python3.8/dist-packages/flask/app.py", line 2344, in process_response
    self.session_interface.save_session(self, ctx.session, response)
  File "/usr/local/lib/python3.8/dist-packages/airflow/www/session.py", line 33, in save_session
    return super().save_session(*args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/flask_session/sessions.py", line 573, in save_session
    response.set_cookie(app.config["SESSION_COOKIE_NAME"], session_id,
  File "/usr/local/lib/python3.8/dist-packages/werkzeug/sansio/response.py", line 224, in set_cookie
    dump_cookie(
  File "/usr/local/lib/python3.8/dist-packages/werkzeug/http.py", line 1303, in dump_cookie
    if not _cookie_no_quote_re.fullmatch(value):
TypeError: cannot use a string pattern on a bytes-like object

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/flask/app.py", line 1847, in finalize_request
    response = self.process_response(response)
  File "/usr/local/lib/python3.8/dist-packages/flask/app.py", line 2344, in process_response
    self.session_interface.save_session(self, ctx.session, response)
  File "/usr/local/lib/python3.8/dist-packages/airflow/www/session.py", line 33, in save_session
    return super().save_session(*args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/flask_session/sessions.py", line 573, in save_session
    response.set_cookie(app.config["SESSION_COOKIE_NAME"], session_id,
  File "/usr/local/lib/python3.8/dist-packages/werkzeug/sansio/response.py", line 224, in set_cookie
    dump_cookie(
  File "/usr/local/lib/python3.8/dist-packages/werkzeug/http.py", line 1303, in dump_cookie
    if not _cookie_no_quote_re.fullmatch(value):
TypeError: cannot use a string pattern on a bytes-like object
90.3.138.42 - - [01/Nov/2023:14:24:42 +0000] "GET /favicon.ico HTTP/1.1" 500 1539 "http://fickenwald.openabal.com/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/118.0.0.0 Safari/537.36"



### What you think should happen instead

I think it is quite obvious why I think this is wrong. It simply does not work.

### How to reproduce

# update the ubuntu version
apt-get --yes update
apt-get --yes install net-tools

# install the python and pip stuff
apt-get --yes install python3-pip
pip3 install --upgrade pip==20.2.4

# install airflow
export AIRFLOW_HOME=~/airflow
pip3 install apache-airflow

# initialise airflow
airflow db init
airflow users create --username admin --firstname jamie  --lastname marshall --role Admin --email ijm@amenesik.com --password=password

# launch airflow web server
airflow webserver --port 80 > airflow.log 2> airflow.err &

# launch airflow scheduler
airflow scheduler > scheduler.out 2> scheduler.err &
```

### Operating System

NAME="Ubuntu" VERSION="20.04.2 LTS (Focal Fossa)" ID=ubuntu ID_LIKE=debian PRETTY_NAME="Ubuntu 20.04.2 LTS" VERSION_ID="20.04" HOME_URL="https://www.ubuntu.com/" SUPPORT_URL="https://help.ubuntu.com/" BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/" PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy" VERSION_CODENAME=focal UBUNTU_CODENAME=focal

### Versions of Apache Airflow Providers

so far i am not running any DAGS it is a fresh vanilla installation
I try and connect to AIRFLOW and CRASH.

### Deployment

Other

### Anything else

ALWAYS : very frustrating

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template! If you are willing to raise PR to address this issue please do so, no need to wait for approval.

**Comment by Taragolis** [comment]
Looks like an installation issue, I would recommend to follow recommendation from [Installation from PyPI](https://airflow.apache.org/docs/apache-airflow/stable/installation/installing-from-pypi.html#installation-from-pypi) first. It might involve some broken python packages dependencies into your system, look like you install as a root user

**Comment by ijm667** [comment]
So I followed the rcommendations:
  Attempting uninstall: PyYAML
    Found existing installation: PyYAML 5.3.1
ERROR: Cannot uninstall 'PyYAML'. It is a distutils installed project and thus we cannot accurately determine which files belong to it which would lead to only a partial uninstall.

**Reply by Taragolis** [reply]
Are you tried to install Airflow within virtual environment?

**Comment by ijm667** [comment]
it would seem that airflow is no longer compatible with Ubuntu 20.04

**Reply by potiuk** [reply]
> it would seem that airflow is no longer compatible with Ubuntu 20.04

Not really. You simply are not aware how it works with distribution based Python installation. 

For installing applications that are not managed by your distribution (via `apt` in uour case) you should always follow recommendations from PyPI and Python maintainers to always install "standard" python rather than the one that distribution maintainers trim and put as "managed by distribution". 

There is even PEP for that (which has links to more information https://peps.python.org/pep-0668/) to improve experience of users like you who face similar issues.  You can follow the links there to learn more.

The problem is that often Python that comes with distribution is heavily modified and incompatible with "standard" python and some packages come pre-installed and some are managed externally (that's the PEP about) - and this is very messy if you try to combine `pip` with those distro packages. Everyone in the community including pip maintainers ALWAYS recomend you to install stock python rather than this coming with distro of yours if you want  install anything else than what is coming with `apt` installation on the distribution. 

Here is what you should (always) do when you install Python application outside of the distribution inventory (`apt` in your case):

1) Install python using mechanisms recommended by Python communty (I personally recommend pyenv)
2) Set it up as a global Python (see pyenv docs for that)
2) Create venv using that stock Python
3) Install airflow in this venv

This is a general knowledge of everyone who uses Python for anything "serious".  You wrote "I followed the rcommendations:".
But maybe we demand a bit too much knowledge from our users?  

So I have a very kind request after you try the above and get it working.

Since you got help and advise here how to fix the problem, can you please submit PR with  update to our documentation explaining what I just advised you - in the words and place that users like you will be able to find and follow ?

It's very easy - just go to the page that @Taragolis helpfully pointed out to you and click "Suggest a change on this page" and PR will be opened for you. - then you can describe it - straight in the GitHub UI - directly in the docs sources and submit the PR for review by maintainers.

I think - even if this is a genarl knowledges, somehow users like you do not understand that this is the general practice. So it would be great if you could help users like you who undoubtedly when they hit problem like this they will be looking for help in our docs, so I think you might be the best person to know where you looked and where this information might be helpfiul to add and explain.

Can we count on your help @ijm667  if the advise I gave you helps you ? Would you like to help other users like you ?

**Comment by ijm667** [comment]
Hello Jarek,

I read your email last night, late, and had the intention of looking into the PR today and complying with your very reasonable request.

I had managed to find a formula that worked yesterday
 (
    pip install "apache-airflow==2.7.2" apache-airflow-providers-google==10.1.1
    airflow db init
    airflow create user ...
    airflow webserver ... &
    airflow scheduler ... &
)

and had managed to connect to the airflow webserver using google chrome and the created user credentials.
I packaged this all up in my deployment script, dropped the deployed VM, relanched the entire process to check, and all was good.

Today, I had a lot of other things to handle, work especially, then this evening I relanched the entire process and encountered an totally different series of problems to resolve.


  *   the pip install throws a load of conflicts and things missing

  *   the airflow db init complains that it is deprectaed and i should use airflow db migrate

  *   airflow db migrate crashed with the traditional python stack trace of complaints

  *   naturally airflow webserver and airflow scheduler were equally unhappy.

This really is a problem.

I don't have the time to debug this product for you, I already have to cope with using a sanitized version of Airflow, with my major customer, I dread the day that they will decide to change version, it will probably require another long period before it reaches stability.

I think the thing that sums it up best is a quotation from the Airfow Documentation concerning the SimpleHttpOperator :

"For historical reasons, configuring HTTPS connectivity via HTTP operator is, well, difficult and counter-intuitive. "

and

"As counter-intuitive as it is, this is historically the way how the operator/hook works and it’s not easy to change without breaking backwards compatibility because there are other operators build on top of the SimpleHttpOperator that rely on that functionality and there are many users using it already."

I have been working in the field of professional software development for over thirty years now and it should be so easy to offer a BetterHttpOperator that provides a more logical approach to the subject of HTTP operator connections, leaving the simple version for those that are stuck with it.

I do understand that the Python community has its reasons, as does the Java community, as does the Scala community, the Go community, the Erlang community etc ...

I think I will lose less time writing the schedular code for my use case, it is something that has been niggling me now after living with my customer's Airflow idiosynchronicities over these past months, so if and when time permits ....

Sincerely and Respectfully
Jamie Marshall

________________________________

**Reply by Taragolis** [reply]
When you install such a big project as Airflow without constraints file it is a good chance that you might have something bad.
It might happen because new version of dependency released, or dependency of dependency  released, or dependency of dependency of dependency, or ... i guess you get my point here.

Even if you ignore all recommendation and documentation I would provide this simple code which would work on most modern Posix system (not only for airflow, but I guess Airflow one of the project which provide constraints files): 

```bash
# Step 1: Move to required directore, lets assume it is user home directory
cd ~

# Step 2: Create new vistual environment, might required to install Ubuntu package python3-venv
python -m venv airflow-venv

# Step 3: Activate it, so we could use it
source airflow-venv/bin/activate

# Step 4: Retrieve path to python interpreter from venv
# E.g. in my case it is /Users/taragolis/airflow-venv/bin/python (MacOS)
which python

# Step 5: Prepare variables
AIRFLOW_VERSION=2.7.2
PYTHON_VERSION="$(python --version | cut -d " " -f 2 | cut -d "." -f 1-2)
CONSTRAINT_URL="https://raw.githubusercontent.com/apache/airflow/constraints-${AIRFLOW_VERSION}/constraints-${PYTHON_VERSION}.txt"

# Step 6: Update pip in venv
pip install -U pip

# Step 7: Install Airflow and extras
pip install "apache-airflow[google]==${AIRFLOW_VERSION}" --constraint "${CONSTRAINT_URL}"

# Step 8: Create installed packages and save it, might be required for repeatable installs
# And could be used instead of Step 7, by call in fresh (!!!) virtual environemnt with the same (!!!) minor version of python
# pip install -r 
pip freeze > my-awesome-requirements.txt

# Step 9: Deactivate Virtual environment (optional)
deactivate

# Step 10: Run Airflow commands
# This required path to virtual environment python interpreter from Step 4
here-path-to-interpreter-from-venv -m airflow version
```

Most of the information it is basic copy paste from Airflow Documentation, and other just a generic stuff of [python virtual environments](https://docs.python.org/3/library/venv.html) and [pip](https://pip.pypa.io/en/stable/)

**Reply by potiuk** [reply]
Yes @ijm667 - I think when you work in industry for a long time you tend to not follow the instructions and recommendations and think that you know better. But I heartily recommend to follow them, especially that we spent a ton of time to make it easy when you, well actually DO follow the instructions.

Our instructions clearly tell you that THE ONLY WAY of installing airflow reproducibly is to use constraints:

https://airflow.apache.org/docs/apache-airflow/stable/installation/installing-from-pypi.html

And then you have those ready recipes for upgrading your dependencies after you install airflow reproducibly

https://airflow.apache.org/docs/apache-airflow/stable/installation/installing-from-pypi.html#upgrading-and-installing-dependencies-including-providers

Those instructions clearly explain what you should do, yet you failed to follow them.

It's you who clarly ignored those instructions and run a command to install Airflow the way we described, and got into (likely) problem with connexion (relased yesterday) which broke clean installation from scratch if you DO NOT use constratints. Which precisely shows why following the instructions is a good idea. 

Shooting yourself in a foot is not a good idea, yet if you do it on your own - don't blame others.

Also if you are interested why it works this way and why not following the instructions we give you is a bad idea you can watch my talk from the Airflow Summit describing it: 

https://www.youtube.com/watch?v=OhS-RR0gKEU

As a seasoned professional I think you will appreciate that we went to a great lenght not only solving the dependency problem for our users by providing instructions to follow, and also a lot of effort to explain to professionals like you who undoubtedly would like to understand why the process is more complex than usual - even in the form of recorded and published talks.

So as a summary - I heartily recommend you to weight your words, and avoid making judgments of someone being at fault, when it's you who have not followed the guidelines and instructions.

It's simply, and humanly unfair to throw accusations at others, when the fault of negligence of following good instructions is on your side. 

Sincerely,

**Comment by zubairx75** [comment]
can we edit [images ](https://carxapk.net/best-racing-games-for-android/)
